pack/nodejs/helm.ts

```diff
@@ -0,0 +1,715 @@
+    // that would be retrieved in-cluster would be assigned fake values, and none of Tiller's
+    // server-side validity testing is executed.
+    //
+    // Important caveat: Charts often assign hard-coded names for objects; since Pulumi is not able
+    // to assign such names, any time a resource must be replaced, it will be marked as
+    // delete-before-replace. For example, in a normal Pulumi program, a change in a ConfigMap's
+    // would result in: (1) a new ConfigMap being created, (2) all Deployments referencing that
+    // ConfigMap being updated to refer to the new ConfigMap, and (3) deleting the old ConfigMap
+    // only after all the rollouts of those Deployments have succeeded. In contrast, since the
+    // chart asks for a specific name for this ConfigMap, it must be deleted before being replaced
+    // with a ConfigMap of the same name.
+    export class Chart extends pulumi.ComponentResource {
+        public readonly resources: pulumi.CustomResource[];
+
+        constructor(releaseName: string, config: ChartOpts, opts?: pulumi.ResourceOptions) {
+            super("kubernetes:helm.sh/v2:Chart", releaseName, config, opts);
+
+            // Create temporary directories and files to hold chart data and override values.
+            const overrides = tmp.fileSync({postfix: ".yaml"});
+            const chartDir = tmp.dirSync({unsafeCleanup: true});
+
+            try {
+                // Fetch chart.
+                fetch(`${config.repo}/${config.chart}`,
+                    {destination: chartDir.name, version: config.version});
+
+                // Create temporary overrides file.
+                const overrides = tmp.fileSync({postfix: ".yaml"});
+                const data = JSON.stringify(config.values || {}, undefined, "  ");
+                fs.writeFileSync(overrides.name, data);
+
+                // Does not require Tiller. From the `helm template` documentation:
+                //
+                // >  Render chart templates locally and display the output.
+                // >
+                // > This does not require Tiller. However, any values that would normally be
+                // > looked up or retrieved in-cluster will be faked locally. Additionally, none
+                // > of the server-side testing of chart validity (e.g. whether an API is supported)
+                // > is done.
+                const chart = `${chartDir.name}/${config.chart}`;
+                const release = shell.quote([releaseName]);
+                const values = shell.quote([overrides.name]);
+                const yamlStream = execSync(
+                    `helm template ${chart} --name ${release} --values ${values}`
+                ).toString();
+                const resourcesObjects = yaml.safeLoadAll(yamlStream);
+                this.resources = fromList(resourcesObjects, config.transformations || [], { ...opts, parent: this });
```

Comment:
Is there value in exposing this as a map from "qualified name" to resource instead of just an array of resources?  That might make it easier for example to look up some specific resource (the exposed `Service`) and pull outputs off of it.

Reply:
I am not sure there is, because:

1. we'd have to create our own way of addressing them (_e.g._, `${group}/${version}/${kind}/${namespace}/${name}`) which would be opaque and unfamiliar to the user, and opaque in the sense that we'd have to decide what to do about exceptions like the `core` namespace.
1. `${name}` itself is pretty unpredictable, since it can be literally anything.

In general instead of having a bag of properties, I think it's useful to expose nice filtering primitives in the `transform` namespace.